app/scheduled/app-client-expire/index.ts

```diff
@@ -0,0 +1,150 @@
+  const db = await tables()
+  const client = db._doc as unknown as DynamoDBDocument
+  const TableName = db.name('client_credentials')
+
+  const deletionDate = new Date()
+  const deletionCutoff = deletionDate.setTime(
+    deletionDate.getTime() - 30 * 86400000
+  )
+
+  const warningDate = new Date()
+  const warningCutoff = warningDate.setTime(
+    warningDate.getTime() - 21 * 86400000
+  )
+
+  const expiredAndWarningCredentials = paginateQuery(
+    {
+      client,
+    },
+    {
+      FilterExpression:
+        'lastUsed < :warningCutoff and attribute_not_exists(expired)',
+      TableName,
+      ExpressionAttributeValues: {
+        ':warningCutoff': warningCutoff,
+      },
+      ExpressionAttributeNames: {
+        '#sub': 'sub',
+      },
+      ProjectionExpression: '#sub, name, client_id, lastUsed',
+    }
+  )
+
+  const expiredCreds: CredentialInfo[] = []
+  const warningCreds: CredentialInfo[] = []
+  for await (const page of expiredAndWarningCredentials) {
+    const items = page.Items as CredentialInfo[]
+    if (items) {
+      expiredCreds.push(
+        ...items.filter(
+          (cred) => !cred.lastUsed || cred.lastUsed <= deletionCutoff
+        )
+      )
+      warningCreds.push(
+        ...items.filter(
+          (cred) => cred.lastUsed && cred.lastUsed <= warningCutoff
+        )
+      )
+    }
+  }
+
+  const expirationDate = Date.now()
+  await Promise.all([
+    // Mark as expired in DynamoDB
+    ...expiredCreds.map((cred) =>
+      db.client_credentials.update({
+        Key: { sub: cred.sub, client_id: cred.client_id },
+        UpdateExpression: 'SET expired = :expired',
+        ExpressionAttributeValues: {
+          ':expired': expirationDate,
+        },
+      })
+    ),
+    // Delete App clients from Cognito
+    ...expiredCreds.map((cred) => deleteAppClient(cred.client_id)),
+  ])
+
+  const usersMapForExpiredClients: { [key: string]: string } =
+    Object.fromEntries(
+      await Promise.all(
+        expiredCreds.map(async (credential) => [
+          extractAttributeRequired(
+            (await getCognitoUserFromSub(credential.sub)).Attributes,
+            'email'
+          ),
+          credential.name,
+        ])
+      )
+    )
+  const expirationEmailPromises = expiredCreds.map((cred) =>
+    sendEmail({
+      to: [usersMapForExpiredClients[cred.name]],
+      fromName: 'GCN Kafka Service',
+      subject: 'Client Credential Deleted',
+      body: `Your Kafka client credential "${cred.name}" has expired. For more information about our credential expiration policy, please visit ${origin}/docs/internal/auth or ${origin}/contact for support.`,
+    })
+  )
+
+  const userMapForWarningClients = Object.fromEntries(
+    await Promise.all(
+      warningCreds.map(async (credential) => [
+        credential.client_id,
+        extractAttributeRequired(
+          (await getCognitoUserFromSub(credential.sub)).Attributes,
+          'email'
+        ),
+      ])
+    )
+  )
```

Comment:
```suggestion
```

Two performance improvements:
- Construct only one mapping from sub to email. You don't need to construct a mapping for users that you are warning _and_ users that are having credentials expire. Some users may be in both categories, and you can save some Cognito API calls by querying their user information only once.
- Construct the _unique_ list of subs before querying Cognito.